src/v/compat/check.h

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "json/document.h"
+#include "json/stringbuffer.h"
+#include "json/writer.h"
+#include "reflection/adl.h"
+#include "seastarx.h"
+#include "serde/serde.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <vector>
+
+namespace compat {
+
+/*
+ * Specialize for types that cannot be copied.
+ */
+template<typename T>
+std::tuple<T, T> compat_copy(T t) {
+    return {t, t};
+}
```

Comment:
probably the purpose is `tuple<serde, adl>`, however I would try to make this clear by either switching to a struct

```
template <typename T>
struct copy {
  T serde, adl;
};
```

or put an explanation in the comment whats the purpose of <0> and <1> of the tuple.

Reply:
Or maybe have a copy function that returns only one `T` copy and `compat_binary` calls this copy function as often as necessary. This would abstract away the details for what the copies are used afterwards - you just implement `T compat_copy(T)`.

Reply:
> probably the purpose is tuple<serde, adl>

no, it's for making progress in the framework invoking the compat checkers when the types can't be copied.

> Or maybe have a copy function that returns only one T copy 

i'm not sure i follow this. if using or making a copy of A destroys A, then we can't progress. maybe i'm missing something.